### PR TITLE
Namespace passing in http header

### PR DESCRIPTION
**Pull request by shopifyski**
Adding alternative to namespace inference from hostname prefix.
Utilizing NAMESPACE_METADATA_KEY to transfer namespace.
Tested with remote_sync example:

* when namespace injected - use namespace from header
* when namespace not injected - use namespace from hostname
cc: @LucioFranco @jeremywrowe @jameswritescode

### CODE REVIEW

File: libsql-server/src/hrana/ws/session.rs [hunk 1]
```diff
@@ -65,60 +88,22 @@ pub enum ResponseError {
...
-        })
-    }
-    // todo dupe #auth
-    let namespace_jwt_key = server
-        .namespaces
-        .with(namespace.clone(), |ns| ns.jwt_key())
-        .await??;
-
-    session.auth = namespace_jwt_key
+    namespace_jwt_key
```
Comment (LucioFranco) [inline libsql-server/src/hrana/ws/session.rs:101]: nice 👍

File: libsql-server/src/http/user/db_factory.rs [hunk 2]
```diff
@@ -46,18 +47,25 @@ pub fn namespace_from_headers(
...
-    let host = headers
-        .get("host")
-        .ok_or_else(|| Error::InvalidHost("missing host header".into()))?
-        .as_bytes();
-    let host_str = std::str::from_utf8(host)
-        .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))?;
+    headers
+        .get(NAMESPACE_METADATA_KEY)
+        .ok_or(Error::InvalidNamespace)
+        .and_then(|h| h.to_str().map_err(|_| Error::InvalidNamespace))
```
Comment (LucioFranco) [inline libsql-server/src/http/user/db_factory.rs:53]: would it make sense to add a `&'static str` to this error variant to specify what failed here? Otherwise we consume the string error here and may be hard to figure out why down the line.

File: libsql/src/replication/client.rs [hunk 3]
```diff
@@ -58,7 +59,12 @@ impl Client {
             .try_into()
             .context("Invalid auth token must be ascii")?;
 
-        let ns = split_namespace(origin.host().unwrap()).unwrap_or_else(|_| "default".to_string());
+
+        let ns = maybe_namespace.unwrap_or_else(||
+            split_namespace(origin.host().unwrap())
+            .unwrap_or_else(|_| "default".to_string())
+        );
```
Comment (LucioFranco) [inline libsql/src/replication/client.rs:66]: this code is slightly hard to read, could we use `else {}` maybe?

**Review by LucioFranco — CHANGES_REQUESTED** [review]
two small nits then lets merge

File: libsql-server/src/hrana/ws/session.rs [hunk 4]
```diff
@@ -22,6 +22,29 @@ pub struct Session {
...
+            bail!(ProtocolError::NotSupported {
+                what: "Repeated hello message",
+                min_version: Version::Hrana2,
+            })
+        }
+        self.auth = auth;
+        Ok(())
+    }
+}
+
```
Comment (haaawk) [inline libsql-server/src/hrana/ws/session.rs:47]: Why is this here? This is already in main -> https://github.com/tursodatabase/libsql/blob/main/libsql-server/src/hrana/ws/session.rs#L26

Could you rebase the change @shopifyski please?

File: libsql-server/src/hrana/ws/conn.rs [hunk 5]
```diff
@@ -206,16 +206,18 @@ async fn handle_client_msg(conn: &mut Conn, client_msg: proto::ClientMsg) -> Res
...
-        None => {
-            session::handle_initial_hello(&conn.server, conn.version, jwt, conn.namespace.clone())
-                .await
-                .map(|session| conn.session = Some(session))
-        }
-        Some(session) => {
-            session::handle_repeated_hello(&conn.server, session, jwt, conn.namespace.clone()).await
-        }
-    };
+    let auth = session::handle_hello(&conn.server, jwt, conn.namespace.clone()).await;
```
Comment (haaawk) [inline libsql-server/src/hrana/ws/conn.rs:209]: This is already in main as well, no? https://github.com/tursodatabase/libsql/blob/main/libsql-server/src/hrana/ws/conn.rs#L209

It generally makes maintainer's work much easier if PRs are rebased and contain only new changes. I would be grateful if you could try to make sure that your PRs are rebased and clean. Thank you!

File: libsql-server/src/hrana/ws/session.rs (hunk 1 above)
Comment (haaawk) [reply libsql-server/src/hrana/ws/session.rs:101]: This is already in main @LucioFranco -> https://github.com/tursodatabase/libsql/blob/main/libsql-server/src/hrana/ws/session.rs#L101

File: libsql-server/src/http/user/db_factory.rs (hunk 2 above)
Comment (haaawk) [inline libsql-server/src/http/user/db_factory.rs:53]: Wouldn't it be easier/cleaner to do:
```
headers
    .get(NAMESPACE_METADATA_KEY)
    .and_then(|h| h.to_str())
    .ok_or(Error::InvalidNamespace)
```
?

Comment (haaawk) [reply libsql-server/src/http/user/db_factory.rs:53]: I don't understand this whole block of code though. It contains errors but at the end it has `.or_else` which ignores those errors:

```
headers
        .get(NAMESPACE_METADATA_KEY)
        .ok_or(Error::InvalidNamespace) // <=== 1. translating any error to InvalidNamespace
        .and_then(|h| h.to_str().map_err(|_| Error::InvalidNamespace)) // <=== 2. possibly InvalidNamespace if header value can't be translated to string
        .and_then(|n| NamespaceName::from_string(n.into()))
        .or_else(|_| { // <=== ignoring both error from (1) and from (2)
            let host = headers
                .get("host")
                .ok_or_else(|| Error::InvalidHost("missing host header".into()))?
                .as_bytes();
            let host_str = std::str::from_utf8(host)
                .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))?;

            match split_namespace(host_str) {
                Ok(ns) => Ok(ns),
                Err(_) if !disable_default_namespace => Ok(NamespaceName::default()),
                Err(e) => Err(e),
            }
        })
```
What do I miss? How are `.ok_or(Error::InvalidNamespace)` and `.map_err(|_| Error::InvalidNamespace)` changing anything if the error gets ignore at the end by `.or_else(|_| {` step?

File: libsql/src/replication/client.rs (hunk 3 above)
Comment (shopifyski) [reply libsql/src/replication/client.rs]: you're right, rewritten to more expressive form

File: libsql/src/database/builder.rs [hunk 6]
```diff
@@ -226,6 +228,11 @@ cfg_replication! {
 
         }
 
+        pub fn namespace(mut self, namespace: &str) -> Builder<RemoteReplica> {
```
[the remark this answers is not in the document]
Comment (shopifyski) [reply libsql/src/database/builder.rs]: lemme know if that's better

File: libsql-server/src/http/user/db_factory.rs (hunk 2 above)
Comment (shopifyski) [reply libsql-server/src/http/user/db_factory.rs:53]: yea, this code needs to change
need to define explicit rule of when to use namespace from header and when from hostname
@LucioFranco @haaawk 
what do you think about this?
* If get(NAMESPACE_METADATA_KEY) gives Some, use the header.
* Do not fallback to namespace from hostname in case of malformed header, just return Error
* If get(NAMESPACE_METADATA_KEY) gives None, use namespace from hostname
* If disable_default_namespace not set and any of the above fails, return default namespace
* if disable_default_namespace is set, any error produced by the above should be returned to caller

Comment (shopifyski) [reply libsql-server/src/http/user/db_factory.rs:53]: committed a new proposal

File: libsql/src/database/builder.rs [hunk 7]
```diff
@@ -226,6 +228,12 @@ cfg_replication! {
 
         }
 
+        pub fn namespace<T>(mut self, namespace: T) -> Builder<RemoteReplica>
```
Comment (haaawk) [inline libsql/src/database/builder.rs:231]: Would `pub fn namespace(mut self, namespace: impl Into<String>) -> Builder<RemoteReplica>` do the same?

File: libsql/src/replication/client.rs [hunk 8]
```diff
@@ -58,7 +59,15 @@ impl Client {
             .try_into()
             .context("Invalid auth token must be ascii")?;
 
-        let ns = split_namespace(origin.host().unwrap()).unwrap_or_else(|_| "default".to_string());
+        let ns = if let Some(ns_from_arg) = maybe_namespace
+        {
```
Comment (haaawk) [inline libsql/src/replication/client.rs:63]: Nit: this `{` should probably be at the end of the previous line

File: libsql-server/src/http/user/db_factory.rs (hunk 2 above)
Comment (haaawk) [reply libsql-server/src/http/user/db_factory.rs:53]: The proposal sounds good to me. You said that you committed it but I don't see code in `namespace_from_headers` being changed, is it?

File: libsql/src/replication/client.rs (hunk 8 above)
Comment (shopifyski) [reply libsql/src/replication/client.rs:63]: could that be configured so cargo fmt fixes it automatically?

Comment (haaawk) [reply libsql/src/replication/client.rs]: I would hope cargo fmt to fix it by default.

File: libsql/src/database/builder.rs (hunk 7 above)
Comment (shopifyski) [reply libsql/src/database/builder.rs]: it wouldn't compile afaik

Comment (shopifyski) [reply libsql/src/database/builder.rs]: ur right, fixed

File: libsql-server/src/http/user/db_factory.rs (hunk 2 above)
Comment (shopifyski) [reply libsql-server/src/http/user/db_factory.rs]: pushed

File: libsql-server/src/http/user/db_factory.rs [hunk 9]
```diff
@@ -46,17 +47,27 @@ pub fn namespace_from_headers(
...
+        headers
+            .get("host")
+            .ok_or_else(|| Error::InvalidHost("missing host header".into()))
+            .and_then(|res| Ok(res.as_bytes()))
+            .and_then(|host| {
+                std::str::from_utf8(host)
+                    .map_err(|_| Error::InvalidHost("host header is not valid UTF-8".into()))
+            })
+            .and_then(|host_str| split_namespace(host_str))
+    };
```
Comment (haaawk) [inline libsql-server/src/http/user/db_factory.rs:65]: The behaviour here changed in two ways:
1. If host header is missing then previously we were always returning Err. New version returns Err only if disbale_default_namespace is true
2. If std::str::from_utf8(host) fails then previously we were always returning Err. New version returns Err only if disable_default_namespace is true

I think it is safer to keep the old behaviour. New behaviour is covering error cases like host error being set to wrong value.

Comment (shopifyski) [reply libsql-server/src/http/user/db_factory.rs:65]: I agree with it being safer not to fall back to default so aggressively
but we now get into an "exception driven" logic territory
e.g. error out when host header non utf, but fallback to default, when host header malformed
is there a reason for this special treatment of that particular error?

Comment (haaawk) [reply libsql-server/src/http/user/db_factory.rs:65]: Let's then fallback to default only when the header is missing and otherwise just error out with appropriate error. It seems safer and is more in line with the logic that was there before.

Comment (shopifyski) [reply libsql-server/src/http/user/db_factory.rs]: @haaawk, I restructured that section for backwards compatibility, let me know if you have further comments

File: libsql/src/database/builder.rs [hunk 10]
```diff
@@ -226,6 +228,12 @@ cfg_replication! {
 
         }
 
+        pub fn namespace(mut self, namespace: impl Into<String>) -> Builder<RemoteReplica>
```
Comment (LucioFranco) [inline libsql/src/database/builder.rs:231]: This needs docs like the other constructors

**Review by LucioFranco — CHANGES_REQUESTED** [review]
LGTM left one thing that needs to be added then we can merge

**Review by LucioFranco — APPROVED** [review]

**Comment by MarinPostma** [comment]
This is a breaking change for the platform. The `x-namespace-bin` should be base64 encoded. We just hit this issue deploying a new canary image.

FYI, I'll patch that to base64 decode the header value.
